File: 001TwoSum/001TwoSum.py
#·¨1
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nums_length=len(nums)
        for i in range(nums_length):
            for j in range(i+1,nums_length):
                if(nums[j]==target-nums[i]):
                    return [i,j]
        
#·¨2
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash_map = {}
        for ind, val in enumerate(nums):
            hash_map[val] = ind

        for ind1, val in enumerate(nums):
	    compent = target-val
            if compent in hash_map:
                ind2 = hash_map[compent]
                if ind1!=ind2:
                    return ind1, ind2

if __name__=="__main__":
    print Solution().twoSum([3, 2, 4], 6)

#·¨3
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        hash_map = {}
        for ind, val in enumerate(nums):
            complement=target-val
            if complement in hash_map:
                return  hash_map[complement],ind
            hash_map[val] = ind
if __name__=="__main__":
    print Solution().twoSum([3, 2, 4], 6)

File: 001TwoSum/001TwoSum.java
public class Solution {
//方法1：Time complexity:O(n2)   Space complexity:O(1)
//    public int[] twoSum(int[] nums, int target) {
//        for(int i=0;i<nums.length;i++)
//          for(int j=i+1;j<nums.length;j++)
//                if((nums[j]==target-nums[j]))
//                    return new int[]{i, j};
//    throw new IllegalArgumentException("No two sum solution");    
//    }

//方法2：Time complexity:O(n)   Space complexity:O(n)
public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
        map.put(nums[i], i);
    }
    for (int i = 0; i < nums.length; i++) {
        int complement = target - nums[i];
        if (map.containsKey(complement) && map.get(complement) != i) {
            return new int[] { i, map.get(complement) };
        }
    }
    throw new IllegalArgumentException("No two sum solution");
}

//方法3：Time complexity:O(n)   Space complexity:O(n)
public int[] twoSum(int[] nums, int target) {
    Map<Integer, Integer> map = new HashMap<>();
    for (int i = 0; i < nums.length; i++) {
        int complement = target - nums[i];
        if (map.containsKey(complement)) {
            return new int[] { map.get(complement), i };
        }
        map.put(nums[i], i);
    }
    throw new IllegalArgumentException("No two sum solution");
}
